File: 2ºCurso/PrimerCuatri/PDOO/Practicas/P1/Ejercicios/EjerciciosJava/src/Parcela.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package ejercicios;
/**
 *
 * @author ylope
 */
public class Parcela {
    private static float FACTORALQUILERCALLE = 1.0f;
    private static float FACTORALQUILERCASA = 1.0f;
    private static float FACTORALQUILERHOTEL = 4.0f;
    private String nombre;
    private float precioCompra, precioEdificar, precioBaseAlquiler;
    private int numCasas, numHoteles;
    
    //Constructor Publico
    public Parcela (String n, float pCompra, float pEdificar, float pBaseAlquiler){
        nombre = n;
        precioCompra = pCompra;
        precioEdificar = pEdificar;
        precioBaseAlquiler = pBaseAlquiler;
        numCasas = 0;
        numHoteles = 0;
    }
    
    void Detalles(){
        System.out.println("Parcela: " +nombre
                            + "\nPrecio de Compra: " +precioCompra
                            + "\nPrecio Edificar: " +precioEdificar
                            + "\nPrecio Base de Alquiler: " +precioBaseAlquiler
                            + "\nNumero de casas: " +numCasas
                            + "\nNumero de hoteles: " +numHoteles
                            + "\nPrecio de Alquiler Completo:" +getPrecioAlquilerCompleto() +"\n\n"
        );
    }

    public String getNombre() {
        return nombre;
    }

    public float getPrecioCompra() {
        return precioCompra;
    }

    public float getPrecioEdificar() {
        return precioEdificar;
    }

    public int getNumCasas() {
        return numCasas;
    }

    public int getNumHoteles() {
        return numHoteles;
    }
    
    public float getPrecioAlquilerCompleto(){
        return precioBaseAlquiler*(FACTORALQUILERCALLE + numCasas*FACTORALQUILERCASA + numHoteles*FACTORALQUILERHOTEL);
    }
 
    public boolean construirCasa(){
        numCasas++;
        return true;
    }
    
    public boolean construirHotel(){
        numHoteles++;
        return true;
    }
    
    public boolean igualdadIdentidad (Parcela otraParcela){
        return otraParcela == this;
    }
    public boolean igualdadEstado (Parcela otraParcela){
        return nombre.equals(otraParcela.getNombre())
            && precioCompra == otraParcela.getPrecioCompra()
            && precioEdificar == otraParcela.getPrecioEdificar()
            && getPrecioAlquilerCompleto() == otraParcela.getPrecioAlquilerCompleto()
            && numCasas == otraParcela.getNumCasas()
            && numHoteles == otraParcela.getNumHoteles();
    }
}

File: 2ºCurso/PrimerCuatri/PDOO/Practicas/P1/Ejercicios/EjerciciosRuby/Parcela.rb
class Parcela
  #Variables de clase
  # @@FACTORALQUILERCALLE = 1.0
  # @@FACTORALQUILERCASA = 1.0
  # @@FACTORALQUILERHOTEL = 4.0
  @FACTORALQUILERCALLE = 1.0
  @FACTORALQUILERCASA = 1.0
  @FACTORALQUILERHOTEL = 4.0
  def self.factor_alquiler_calle
    return @FACTORALQUILERCALLE
  end

  def self.factor_alquiler_casa
    return @FACTORALQUILERCASA
  end

  def self.factor_alquiler_hotel
    return @FACTORALQUILERHOTEL
  end

  def initialize(n, p_compra, p_edificar, p_base_alquiler) #constructor
    @nombre = n
    @precio_compra = p_compra
    @precio_edificar = p_edificar
    @precio_base_alquiler = p_base_alquiler
    @numCasas = 0
    @numHoteles = 0
  end

  public
  def Detalles #Imprime los datos de la clase
    puts "Nombre de la parcela: #@nombre"
    puts "Precio de compra: #@precio_compra"
    puts "Precio de edificación: #@precio_edificar"
    puts "Precio base de alquiler: #@precio_base_alquiler"
    puts "Numero de casas: #@numCasas"
    puts "Numero de Hoteles: #@numHoteles"
    puts "El Precio total a pagar es: #{get_precio_alquiler_completo}\n\n"
  end
  #Metodos get de las variables de la clase
  attr_reader :nombre
  attr_reader :precio_compra
  attr_reader :precio_edificar
  attr_reader :precio_base_alquiler

  def get_precio_alquiler_completo
    return @precio_base_alquiler * (Parcela.factor_alquiler_calle +
                                   @numCasas * Parcela.factor_alquiler_casa +
                                   @numHoteles * Parcela.factor_alquiler_hotel)
  end

  #2 metodos bool que aumentan en 1 la construccion designada
  def construir_casa?
    @numCasas+=1
    return true
  end
  def construir_hotel?
    @numHoteles+=1
    return true
  end
end